File: src/compiler/lexer/token_stream.rs
use crate::{
    error::kaori_error::{self, KaoriError},
    kaori_error,
};

use super::{span::Span, token::Token, token_kind::TokenKind};

#[derive(Debug, Clone)]
pub struct TokenStream {
    source: String,
    tokens: Vec<Token>,
    index: usize,
}

impl TokenStream {
    pub fn new(source: String, tokens: Vec<Token>) -> Self {
        Self {
            source,
            tokens,
            index: 0,
        }
    }

    pub fn at_end(&mut self) -> bool {
        return self.token_kind() == TokenKind::EndOfFile;
    }

    pub fn advance(&mut self) {
        self.index += 1;
    }

    pub fn consume(&mut self, expected: TokenKind) -> Result<(), KaoriError> {
        let found = self.token_kind();

        if expected == found {
            self.advance();
            return Ok(());
        } else {
            let span = self.span();

            return Err(kaori_error!(
                span,
                "expected {:?}, but found {:?}",
                expected,
                found,
            ));
        }
    }

    pub fn look_ahead(&mut self, expected: &[TokenKind]) -> bool {
        for i in 0..expected.len() {
            let j = self.index + i;

            if j >= self.tokens.len() {
                return false;
            }

            if self.tokens[j].kind == expected[i] {
                continue;
            }

            return false;
        }

        return true;
    }

    pub fn token_kind(&mut self) -> TokenKind {
        let token = self.tokens.get(self.index).unwrap();

        return token.kind.clone();
    }

    pub fn span(&mut self) -> Span {
        return self.tokens[self.index].span;
    }

    pub fn lexeme(&mut self) -> String {
        let span = self.span();

        return self.source[span.start..span.end].to_string();
    }
}

File: legacy-java/src/main/java/com/kaori/compiler/lexer/TokenStream.java
package com.kaori.compiler.lexer;

import java.util.List;

import com.kaori.error.KaoriError;

public class TokenStream {
    private final List<Token> tokens;
    private String source;
    private int index;
    private int line;

    public TokenStream(List<Token> tokens, String source) {
        this.tokens = tokens;
        this.source = source;
        this.index = 0;
        this.line = 1;
    }

    public TokenKind getCurrent() {
        if (this.atEnd()) {
            return TokenKind.EOF;
        }

        Token token = this.tokens.get(this.index);

        return token.kind();
    }

    public int getLine() {
        Token token = this.tokens.get(this.index);

        return token.line();
    }

    private int getPosition() {
        Token token = this.tokens.get(this.index);

        return token.position();
    }

    private int getSize() {
        Token token = this.tokens.get(this.index);

        return token.size();
    }

    public String getLexeme() {
        return this.source.substring(this.getPosition(), this.getPosition() + this.getSize());
    }

    public boolean atEnd() {
        return this.index >= this.tokens.size();
    }

    private void advance() {
        this.index++;

        if (!this.atEnd()) {
            this.line = this.getLine();
        }
    }

    public void consume(TokenKind expected) {
        if (this.getCurrent() == TokenKind.EOF) {
            throw KaoriError.SyntaxError(
                    String.format("expected %s, but had an unexpected end of file", expected),
                    this.line);
        }

        if (this.getCurrent() != expected) {
            throw KaoriError.SyntaxError(String.format("expected %s instead of %s", expected, this.getCurrent()),
                    this.line);
        }

        this.advance();
    }

    public void consume() {
        this.advance();
    }

    public boolean lookAhead(TokenKind... expected) {
        for (int i = 0; i < expected.length; i++) {
            int j = this.index + i;

            if (j >= this.tokens.size()) {
                return false;
            }

            Token token = this.tokens.get(j);

            if (token.kind() != expected[i])
                return false;
        }

        return true;
    }
}
